services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedTenantService.java

```diff
@@ -0,0 +1,528 @@
+                        .map(updated -> {
+                            span.log(String.format("successfully updated tenant [%s]", tenantId));
+                            return Future.<OperationResult<Void>> succeededFuture(OperationResult.ok(
+                                    HttpURLConnection.HTTP_NO_CONTENT,
+                                    null,
+                                    Optional.empty(),
+                                    Optional.of(updateResult.getString(MongoDbDeviceRegistryUtils.FIELD_VERSION))));
+                        })
+                        .orElse(MongoDbDeviceRegistryUtils.checkForVersionMismatchAndFail(
+                                tenantId, resourceVersion, findTenant(tenantId))))
+                .recover((error) -> {
+                    if (error instanceof MongoCommandException) {
+                        logDuplicationError(((MongoCommandException) error).getErrorMessage(), tenantId, span);
+                        return Future.succeededFuture(OperationResult.empty(HttpURLConnection.HTTP_CONFLICT));
+                    } else if (error instanceof ServiceInvocationException) {
+                        return Future.failedFuture(((ServiceInvocationException) error));
+                    }
+                    return Future.failedFuture(new ServerErrorException(HttpURLConnection.HTTP_INTERNAL_ERROR,
+                            String.format("Error modifying tenant [%s].", tenantId)));
+                });
+    }
+
+//
+    /**
+     * {@inheritDoc}
+     * @return A future indicating the outcome of the operation.
+     *         The <em>status</em> will be
+     *         <ul>
+     *         <li><em>204 No Content</em> if the tenant has been removed successfully.</li>
+     *         <li><em>403 forbidden</em> if tenant modification is forbidden due to configuration settings.</li>
+     *         <li><em>404 Not Found</em> if no tenant with the given identifier and version exists.</li>
+     *         <li><em>412 Precondition Failed</em> if the version exists but mismatch the current tenants version.</li>
+     *         </ul>
+     */
+    @Override
+    public Future<Result<Void>> deleteTenant(final String tenantId, final Optional<String> resourceVersion,
+            final Span span) {
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(resourceVersion);
+        Objects.requireNonNull(span);
+
+        return MongoDbDeviceRegistryUtils.isAllowedToModify(config, tenantId)
+                .compose(ok -> result(tenantId, resourceVersion, span))
+                .recover(error -> Future.succeededFuture(MongoDbDeviceRegistryUtils.mapErrorToResult(error, span)));
+    }
+
+    private Future<Result<Void>> result(final String tenantId, final Optional<String> resourceVersion,
+            final Span span) {
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(resourceVersion);
+        Objects.requireNonNull(span);
```

Comment:
Redundant null checks in the private methods could be removed for those parameters that have been already checked.

Reply:
done